compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -67,7 +69,14 @@ struct ObligationStorage<'tcx> {
 }
 
 impl<'tcx> ObligationStorage<'tcx> {
-    fn register(
+    fn register(&mut self, obligation: PredicateObligation<'tcx>) {
+        if self.dedup.insert((obligation.param_env, obligation.predicate)) {
```

Comment:
this dedups stuff with different recursion depths which seems wrong. in general this makes me nervous that we'll have a sort of "cache" that doesnt take into account all the inputs. Why not `FxHashSet<PredicateObligation<'tcx>>` which is the exact type used for the list of obligations to prove?